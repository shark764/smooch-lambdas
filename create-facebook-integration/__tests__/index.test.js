/* eslint-disable max-len */
const { lambda: { api: { validateTenantPermissions } } } = require('alonzo');
const axios = require('axios');

jest.mock('alonzo');
jest.mock('serenova-js-utils/strings');
jest.mock('axios');

validateTenantPermissions.mockReturnValue(true);
global.Date.prototype.toISOString = jest.fn(() => 'January 1 1970');

const mockGetSecretValue = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        '5e31c81640a22c000f5d7f28-id': 'id',
        '5e31c81640a22c000f5d7f28-secret': 'secret',
      }),
    }),
  }));


const mockCreateIntegration = jest.fn()
  .mockImplementation(() => ({
    integration: {
      id: '667802d812321342',
      type: 'messenger',
      displayName: 'display-name',
      pageName: 'page-name',
    },
  }));

const mockUpdate = jest.fn(() => ({}))
  .mockImplementation(() => ({
    promise: () => ({
      Attributes: {
        type: 'facebook',
        status: 'done',
      },
    }),
  }));

axios.mockImplementation(() => Promise.resolve({
  status: 200,
  data: {
    access_token: 'access-token',
  },
}));

const mockBody = {
  appId: '5e31c81640a22c000f5d7f28',
  name: 'smooch',
  facebookAppId: '5e31c81640a22c000f5d7f27',
  facebookAppSecret: '123457',
  facebookPageId: '12345678',
  facebookUserAccessToken: 'short-user-access-token',
  description: 'description',
  clientDisconnectMinutes: 150,
};

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
  body: mockBody,
};

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn(),
  },
  SecretsManager: jest.fn().mockImplementation(() => ({ getSecretValue: mockGetSecretValue })),
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({ update: mockUpdate })),
  },
}));

const mockAuthentication = {
  basicAuth: {
    username: 'id',
    password: 'secret',
  },
};

jest.mock('sunshine-conversations-client', () => ({
  ApiClient: {
    instance: {
      authentications: mockAuthentication,
    },
  },
  IntegrationsApi: jest.fn().mockImplementation(() => ({
    createIntegration: mockCreateIntegration,
  })),
  Integration: jest.fn().mockImplementation(() => ({})),
}));

const { handler } = require('../index');

describe('create-faceboook-integration', () => {
  describe('Everthing is successful', () => {
    it('when decription and client disconnect minutes are not provided', async () => {
      mockBody.description = undefined;
      mockBody.clientDisconnectMinutes = undefined;
      const { status } = await handler(event);
      expect(status).toBe(201);
      expect(mockUpdate.mock.calls).toMatchSnapshot();
    });

    it('sends back status 201 when the code runs without error', async () => {
      const result = await handler(event);
      expect(result).toMatchSnapshot();
      expect(axios.mock.calls).toMatchSnapshot();
    });
    describe('Walkthrough', () => {
      beforeAll(async () => {
        jest.clearAllMocks();
        await handler(event);
      });
      it('passes in the correct arguments to validateTenantPermissions', async () => {
        expect(validateTenantPermissions.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to secretsClient.getSecretValue()', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to createIntegration()', async () => {
        expect(mockCreateIntegration.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to docClient.update()', async () => {
        expect(mockUpdate.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to axios', async () => {
        expect(axios.mock.calls).toMatchSnapshot();
      });
    });
  });

  it('passes in the correct arguments to axios when facebook page access token is provided', async () => {
    jest.clearAllMocks();
    const body = {
      appId: '5e31c81640a22c000f5d7f28',
      name: 'smooch',
      facebookAppId: '5e31c81640a22c000f5d7f27',
      facebookAppSecret: '123457',
      facebookPageId: '12345678',
      facebookPageAccessToken: 'page-access-token',
      description: 'description',
      clientDisconnectMinutes: 150,
    };

    const mockEvent = {
      params: {
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
      body,
    };
    await handler(mockEvent);
    expect(axios.mock.calls).toMatchSnapshot();
  });

  it('sends back status 400 when there is a invalid body value', async () => {
    const mockEvent = {
      params: {
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
      body: {

      },
    };
    const result = await handler(mockEvent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when there is a invalid params value', async () => {
    const mockEvent = {
      params: {
        id: '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
      body: mockBody,
    };
    const result = await handler(mockEvent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 403 when there are not enough permissions', async () => {
    validateTenantPermissions.mockReturnValueOnce(false);
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error retrieving digital channels credentials', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error creating facebook integration for tenant', async () => {
    mockCreateIntegration.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error saving records in DynamoDB for tenant', async () => {
    mockUpdate.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when facebook api returns error while long lived user access token', async () => {
    jest.clearAllMocks();
    axios.mockImplementationOnce(new Error());
    mockBody.facebookPageAccessToken = undefined;
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when facebook api returns error while page access token', async () => {
    jest.clearAllMocks();
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(new Error());
    mockBody.facebookPageAccessToken = undefined;
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when fails to parse credentials', async () => {
    jest.clearAllMocks();
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when facebook api for get app token returns error', async () => {
    jest.clearAllMocks();
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when facebook api for delete subscription returns error', async () => {
    jest.clearAllMocks();
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(() => Promise.resolve({
      status: 200,
      data: {
        access_token: 'access-token',
      },
    }));
    axios.mockImplementationOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when both user access token and page access token are missing', async () => {
    mockBody.facebookPageAccessToken = undefined;
    mockBody.facebookUserAccessToken = undefined;
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
