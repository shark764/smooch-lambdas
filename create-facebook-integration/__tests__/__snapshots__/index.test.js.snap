// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-faceboook-integration Everthing is successful Walkthrough passes in the correct arguments to axios 1`] = `
Array [
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?grant_type=fb_exchange_token&client_id=5e31c81640a22c000f5d7f27&client_secret=123457&fb_exchange_token=short-user-access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/12345678?fields=access_token&access_token=access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?client_id=5e31c81640a22c000f5d7f27&client_secret=123457&grant_type=client_credentials",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer access-token",
      },
      "method": "delete",
      "url": "https://graph.facebook.com/v10.0/5e31c81640a22c000f5d7f27/subscriptions?object=page",
    },
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful Walkthrough passes in the correct arguments to createIntegration() 1`] = `
Array [
  Array [
    "5e31c81640a22c000f5d7f28",
    Object {
      "appId": "5e31c81640a22c000f5d7f27",
      "appSecret": "123457",
      "pageAccessToken": "access-token",
      "type": "messenger",
    },
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful Walkthrough passes in the correct arguments to docClient.update() 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#appId": "app-id",
        "#createdBy": "created-by",
        "#name": "name",
        "#type": "type",
        "#updatedBy": "updated-by",
      },
      "ExpressionAttributeValues": Object {
        ":appId": "5e31c81640a22c000f5d7f28",
        ":created": "January 1 1970",
        ":createdBy": "667802d8-2260-436c-958a-2ee0f71f73f0",
        ":name": "smooch",
        ":t": "facebook",
        ":updated": "January 1 1970",
        ":updatedBy": "667802d8-2260-436c-958a-2ee0f71f73f0",
      },
      "Key": Object {
        "id": "667802d812321342",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "ReturnValues": "ALL_NEW",
      "TableName": "us-east-1-dev-smooch",
      "UpdateExpression": "set #type = :t, #appId = :appId, #name = :name, #createdBy = :createdBy, #updatedBy = :updatedBy,
  created = :created, updated = :updated",
    },
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful Walkthrough passes in the correct arguments to secretsClient.getSecretValue() 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful Walkthrough passes in the correct arguments to validateTenantPermissions 1`] = `
Array [
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "FACEBOOK_INTEGRATIONS_APP_UPDATE",
    ],
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful sends back status 201 when the code runs without error 1`] = `
Object {
  "body": Object {
    "result": Object {
      "id": "667802d812321342",
      "pageName": "page-name",
      "type": "messenger",
      "undefined": "done",
    },
  },
  "status": 201,
}
`;

exports[`create-faceboook-integration Everthing is successful sends back status 201 when the code runs without error 2`] = `
Array [
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?grant_type=fb_exchange_token&client_id=5e31c81640a22c000f5d7f27&client_secret=123457&fb_exchange_token=short-user-access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/12345678?fields=access_token&access_token=access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?client_id=5e31c81640a22c000f5d7f27&client_secret=123457&grant_type=client_credentials",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer access-token",
      },
      "method": "delete",
      "url": "https://graph.facebook.com/v10.0/5e31c81640a22c000f5d7f27/subscriptions?object=page",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?grant_type=fb_exchange_token&client_id=5e31c81640a22c000f5d7f27&client_secret=123457&fb_exchange_token=short-user-access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/12345678?fields=access_token&access_token=access-token",
    },
  ],
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?client_id=5e31c81640a22c000f5d7f27&client_secret=123457&grant_type=client_credentials",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer access-token",
      },
      "method": "delete",
      "url": "https://graph.facebook.com/v10.0/5e31c81640a22c000f5d7f27/subscriptions?object=page",
    },
  ],
]
`;

exports[`create-faceboook-integration Everthing is successful when decription and client disconnect minutes are not provided 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#appId": "app-id",
        "#createdBy": "created-by",
        "#name": "name",
        "#type": "type",
        "#updatedBy": "updated-by",
      },
      "ExpressionAttributeValues": Object {
        ":appId": "5e31c81640a22c000f5d7f28",
        ":created": "January 1 1970",
        ":createdBy": "667802d8-2260-436c-958a-2ee0f71f73f0",
        ":name": "smooch",
        ":t": "facebook",
        ":updated": "January 1 1970",
        ":updatedBy": "667802d8-2260-436c-958a-2ee0f71f73f0",
      },
      "Key": Object {
        "id": "667802d812321342",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "ReturnValues": "ALL_NEW",
      "TableName": "us-east-1-dev-smooch",
      "UpdateExpression": "set #type = :t, #appId = :appId, #name = :name, #createdBy = :createdBy, #updatedBy = :updatedBy,
  created = :created, updated = :updated",
    },
  ],
]
`;

exports[`create-faceboook-integration passes in the correct arguments to axios when facebook page access token is provided 1`] = `
Array [
  Array [
    Object {
      "method": "get",
      "url": "https://graph.facebook.com/oauth/access_token?client_id=5e31c81640a22c000f5d7f27&client_secret=123457&grant_type=client_credentials",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer access-token",
      },
      "method": "delete",
      "url": "https://graph.facebook.com/v10.0/5e31c81640a22c000f5d7f27/subscriptions?object=page",
    },
  ],
]
`;

exports[`create-faceboook-integration sends back status 400 when both user access token and page access token are missing 1`] = `
Object {
  "body": Object {
    "error": "Page Access Token and User Access Token missing, Please provide any one of the token",
  },
  "status": 400,
}
`;

exports[`create-faceboook-integration sends back status 400 when there is a invalid body value 1`] = `
Object {
  "body": Object {
    "message": "Error: invalid body value(s). \\"appId\\" is required / \\"facebookAppId\\" is required / \\"facebookAppSecret\\" is required / \\"facebookPageId\\" is required / \\"name\\" is required",
  },
  "status": 400,
}
`;

exports[`create-faceboook-integration sends back status 400 when there is a invalid params value 1`] = `
Object {
  "body": Object {
    "error": [ValidationError: "id" is not allowed],
    "message": "Error: invalid params value(s). \\"id\\" is not allowed",
  },
  "status": 400,
}
`;

exports[`create-faceboook-integration sends back status 403 when there are not enough permissions 1`] = `
Object {
  "body": Object {
    "expectedPermissions": Object {
      "tenant": Array [
        "FACEBOOK_INTEGRATIONS_APP_UPDATE",
      ],
    },
    "message": "Error not enough permissions",
  },
  "status": 403,
}
`;

exports[`create-faceboook-integration sends back status 500 when facebook api for delete subscription returns error 1`] = `
Object {
  "body": Object {
    "error": "Unexpected error occurred deleting current page subscription.",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when facebook api for get app token returns error 1`] = `
Object {
  "body": Object {
    "error": "Unexpected error occurred retrieving App Access token.",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when facebook api returns error while long lived user access token 1`] = `
Object {
  "body": Object {
    "error": "Unexpected error occurred retrieving Page Access token. Please try again with Facebook Login/Another user access token",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when facebook api returns error while page access token 1`] = `
Object {
  "body": Object {
    "error": "Unexpected error occurred retrieving Page Access token. Please try again with Facebook Login/Another user access token",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when fails to parse credentials 1`] = `
Object {
  "body": Object {
    "message": "Failed to parse smooch credentials or credentials does not exists",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when there is a error creating facebook integration for tenant 1`] = `
Object {
  "body": Object {
    "error": [Error],
    "message": "An Error has occurred trying to create a facebook integration for tenant",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when there is a error retrieving digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`create-faceboook-integration sends back status 500 when there is a error saving records in DynamoDB for tenant 1`] = `
Object {
  "body": Object {
    "error": [TypeError: docClient.update(...).promise is not a function],
    "message": "An Error has occurred trying to save a record in DynamoDB for tenant",
  },
  "status": 500,
}
`;
